File: skills/python/assets/cli-main.py
"""The `acme` CLI. Exit codes: 0 ok, 1 some input was invalid, 2 usage error (typer's default).

Commands only parse arguments and format output; every rule lives in acme_core, so the
CLI stays thin and the logic is tested once, in the library.
"""

import json
import sys
from enum import StrEnum
from importlib.metadata import version
from typing import Annotated

import typer

from acme_core import InvalidInputError, new_code, normalize_url, slugify

app = typer.Typer(help="Tools for acme short links.", no_args_is_help=True)


class Format(StrEnum):
    TEXT = "text"
    JSON = "json"


def _version(value: bool) -> None:
    if value:
        typer.echo(f"acme {version('acme-cli')}")
        raise typer.Exit


@app.callback()
def main(
    _version_flag: Annotated[
        bool,
        typer.Option("--version", callback=_version, is_eager=True, help="Show the version."),
    ] = False,
) -> None:
    """Tools for acme short links."""


@app.command()
def check(
    urls: Annotated[
        list[str] | None, typer.Argument(help="URLs to check; reads stdin if none.")
    ] = None,
    fmt: Annotated[Format, typer.Option("--format", "-f", help="Output format.")] = Format.TEXT,
) -> None:
    """Validate and normalize URLs. Exits 1 if any URL is invalid."""
    inputs = urls or [line for line in sys.stdin.read().splitlines() if line.strip()]
    results: list[dict[str, str]] = []
    for raw in inputs:
        try:
            results.append({"input": raw, "url": normalize_url(raw)})
        except InvalidInputError as exc:
            results.append({"input": raw, "error": exc.reason})
    if fmt is Format.JSON:
        typer.echo(json.dumps(results, indent=2))
    else:
        for r in results:
            line = f"ok    {r['url']}" if "url" in r else f"error {r['input']!r}: {r['error']}"
            typer.echo(line, err="error" in r)
    if any("error" in r for r in results):
        raise typer.Exit(code=1)


@app.command()
def code(
    length: Annotated[int, typer.Option(min=3, max=32, help="Code length.")] = 7,
    count: Annotated[int, typer.Option(min=1, max=1000, help="How many codes.")] = 1,
) -> None:
    """Generate random short codes."""
    for _ in range(count):
        typer.echo(new_code(length))


@app.command()
def slug(text: Annotated[list[str], typer.Argument(help="Words to slugify.")]) -> None:
    """Turn free text into a URL slug."""
    result = slugify(" ".join(text))
    if not result:
        typer.echo("error: nothing left to slugify", err=True)
        raise typer.Exit(code=1)
    typer.echo(result)

File: skills/rust/assets/cli-main.rs
//! `acme`: turn text into slugs and ids into short codes (and back).
//!
//! Exit codes: 0 success, 1 runtime error, 2 usage error (clap's default).
use std::io::{self, BufRead, Write};
use std::process::ExitCode;

use acme_core::{ShortCode, Slug};
use anyhow::{Context, Result};
use clap::{Parser, Subcommand};

/// Slugs and short codes for Acme links.
#[derive(Debug, Parser)]
#[command(version, about)]
struct Cli {
    #[command(subcommand)]
    command: Command,
}

#[derive(Debug, Subcommand)]
enum Command {
    /// Normalize text into a URL slug. Reads lines from stdin when no TEXT is given.
    Slug {
        /// Text to slugify.
        text: Option<String>,
    },
    /// Encode a numeric id as a base62 short code.
    Encode {
        /// The id to encode.
        id: u64,
    },
    /// Decode a base62 short code back to its id.
    Decode {
        /// The short code to decode.
        code: String,
    },
}

fn main() -> ExitCode {
    let cli = Cli::parse();
    let stdout = io::stdout();
    match run(cli, &mut stdout.lock(), io::stdin().lock()) {
        Ok(()) => ExitCode::SUCCESS,
        Err(err) => {
            // `{:#}` prints the whole context chain on one line: "decoding \"x\": invalid ..."
            #[expect(clippy::print_stderr, reason = "the CLI's error channel")]
            {
                eprintln!("error: {err:#}");
            }
            ExitCode::FAILURE
        }
    }
}

/// All logic takes its I/O as parameters, so tests drive it without spawning a process.
fn run(cli: Cli, out: &mut impl Write, input: impl BufRead) -> Result<()> {
    match cli.command {
        Command::Slug { text: Some(text) } => {
            writeln!(
                out,
                "{}",
                Slug::parse(&text).with_context(|| format!("slugifying {text:?}"))?
            )?;
        }
        Command::Slug { text: None } => {
            for line in input.lines() {
                let line = line.context("reading stdin")?;
                writeln!(
                    out,
                    "{}",
                    Slug::parse(&line).with_context(|| format!("slugifying {line:?}"))?
                )?;
            }
        }
        Command::Encode { id } => writeln!(out, "{}", ShortCode::from_id(id))?,
        Command::Decode { code } => {
            let parsed: ShortCode = code.parse().with_context(|| format!("decoding {code:?}"))?;
            writeln!(out, "{}", parsed.to_id())?;
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    fn run_args(args: &[&str], stdin: &str) -> Result<String> {
        let cli = Cli::try_parse_from(std::iter::once("acme").chain(args.iter().copied()))?;
        let mut out = Vec::new();
        run(cli, &mut out, stdin.as_bytes())?;
        Ok(String::from_utf8(out)?)
    }

    #[test]
    fn slug_from_arg_and_stdin() {
        assert_eq!(
            run_args(&["slug", "Hello World"], "").unwrap(),
            "hello-world\n"
        );
        assert_eq!(run_args(&["slug"], "A b\nC d\n").unwrap(), "a-b\nc-d\n");
    }

    #[test]
    fn encode_decode() {
        assert_eq!(run_args(&["encode", "125"], "").unwrap(), "21\n");
        assert_eq!(run_args(&["decode", "21"], "").unwrap(), "125\n");
        assert!(run_args(&["decode", "!!"], "").is_err());
    }

    #[test]
    fn clap_definition_is_valid() {
        use clap::CommandFactory;
        Cli::command().debug_assert();
    }
}
